python/ngraph/ops.py

```diff
@@ -633,6 +633,36 @@ def prod(node, reduction_axes=None, name=None):
     return Product(node, AxisSet(get_reduction_axes(node, reduction_axes)))
 
 
+@nameable_op
+def reduce(node,                 # type: Node
+           initial_value,        # type: ScalarData
+           reduction_node,       # type: Node
+           reduction_axes=None,  # type: List[int]
+           name=None,            # type: str
+           ):
+    # type: (...) -> Node
+    """Perform general tensor reduction operation.
+
+    :param node: The node providing data for reduction operation.
+    :param initial_value: The initial value for reduction operation.
+    :param reduction_node: The node providing binary reduction operation. The operation must
+                           accept two nodes providing scalar operands and return a node which
+                           produces a scalar result.
+    :param reduction_axes: The list of axes indices to be reduced. Default to reduce all axes.
+    :param name: The new name for output node.
+    :return: The node performing reduction operation with provided reduction node.
+    """
+    if reduction_axes is None:
+        reduction_axes = list(range(len(node.shape)))
+    init_val_node = constant(initial_value)
+    # wrap reduction node into Function object
+    param1 = Parameter(node.get_element_type(), Shape([]))
+    param2 = Parameter(node.get_element_type(), Shape([]))
+    reduction_operation = Function(NodeVector([reduction_node(param1, param2)]), [param1, param2],
```

Comment:
Let's add an option to accept a `Function` instance directly.

The current code, which takes a Python callable should be one of 2 possible inputs.